mixer/pkg/runtime2/config/ephemeral.go

```diff
@@ -0,0 +1,370 @@
+		log.Debug("Received attribute manifest change event.")
+	}
+
+	switch event.Type {
+	case store.Update:
+		e.entries[event.Key] = event.Value
+	case store.Delete:
+		delete(e.entries, event.Key)
+	}
+}
+
+// BuildSnapshot builds a stable, fully-resolved snapshot view of the configuration.
+func (e *Ephemeral) BuildSnapshot() *Snapshot {
+	id := e.nextID
+	e.nextID++
+
+	log.Debugf("Building new config.Snapshot: id='%d'", id)
+
+	// Allocate new counters, to use with the new snapshot.
+	counters := newCounters(id)
+
+	attributes := e.processAttributeManifests(counters)
+
+	handlers := e.processHandlerConfigs(counters)
+
+	instances := e.processInstanceConfigs(counters)
+
+	rules := e.processRuleConfigs(handlers, instances, counters)
+
+	e.attributesChanged = false
+
+	s := &Snapshot{
+		ID:         id,
+		Templates:  e.templates,
+		Adapters:   e.adapters,
+		Attributes: expr.NewFinder(attributes),
+		Handlers:   handlers,
+		Instances:  instances,
+		Rules:      rules,
+		Counters:   counters,
+	}
+
+	e.cachedAttributes = attributes
+
+	log.Infof("Built new config.Snapshot: id='%d'", id)
+	if log.DebugEnabled() {
+		log.Debugf("config.Snapshot contents:\n%s", s.String())
+	}
+	return s
+}
+
+func (e *Ephemeral) processAttributeManifests(counters Counters) map[string]*config.AttributeManifest_AttributeInfo {
+	if !e.attributesChanged && e.cachedAttributes != nil {
+		return e.cachedAttributes
+	}
+
+	attrs := make(map[string]*config.AttributeManifest_AttributeInfo)
+	for k, obj := range e.entries {
+		if k.Kind != AttributeManifestKind {
+			continue
+		}
+
+		log.Debug("Start processing attributes from changed manifest...")
+
+		cfg := obj.Spec
+		for an, at := range cfg.(*config.AttributeManifest).Attributes {
+			attrs[an] = at
+
+			if log.DebugEnabled() {
+				log.Debugf("Attribute '%s': '%s'.", an, at.ValueType.String())
+			}
+		}
+	}
+
+	// append all the well known attribute vocabulary from the templates.
+	//
+	// ATTRIBUTE_GENERATOR variety templates allows operators to write Attributes
+	// using the $out.<field Name> convention, where $out refers to the output object from the attribute generating adapter.
+	// The list of valid names for a given Template is available in the Template.Info.AttributeManifests object.
+	for _, info := range e.templates {
+		log.Debugf("Processing attributes from template: '%s'", info.Name)
+
+		for _, v := range info.AttributeManifests {
+			for an, at := range v.Attributes {
+				attrs[an] = at
+
+				if log.DebugEnabled() {
+					log.Debugf("Attribute '%s': '%s'", an, at.ValueType.String())
+				}
+			}
+		}
+	}
+
+	log.Debug("Completed processing attributes.")
+	counters.attributes.Add(float64(len(attrs)))
+
+	return attrs
+}
+
+func (e *Ephemeral) processHandlerConfigs(counters Counters) map[string]*Handler {
+	handlers := make(map[string]*Handler)
+
+	for key, resource := range e.entries {
+		var info *adapter.Info
+		var found bool
+		if info, found = e.adapters[key.Kind]; !found {
+			continue
```

Comment:
Could you add a comment here and in similar loops in this file explaining what the !found case means? Why is this happening and why is it OK for it not to be found?